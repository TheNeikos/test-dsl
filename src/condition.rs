use std::marker::PhantomData;

use crate::error::TestErrorCase;

pub trait TestCondition<H>: 'static {
    fn check_now(&self, harness: &H, node: &kdl::KdlNode) -> Result<(), TestErrorCase>;

    fn wait_until_true(&self, harness: &H, node: &kdl::KdlNode) -> Result<(), TestErrorCase>;

    fn clone_box(&self) -> Box<dyn TestCondition<H>>;
}

trait Checker<H>: Clone {
    fn check(&self, harness: &H) -> Option<bool>;
}

impl<H> Checker<H> for () {
    fn check(&self, _harness: &H) -> Option<bool> {
        None
    }
}

impl<T, H> Checker<H> for T
where
    T: Clone,
    T: Fn(&H) -> bool,
{
    fn check(&self, harness: &H) -> Option<bool> {
        Some((self)(harness))
    }
}

pub struct Condition<H, N, W> {
    now: N,
    wait: W,
    _pd: PhantomData<fn(N, W, H)>,
}

impl<H, N> Condition<H, N, ()>
where
    N: Fn(&H) -> bool,
{
    pub fn new_now(now: N) -> Self {
        Condition {
            now,
            wait: (),
            _pd: PhantomData,
        }
    }
}

impl<H, N: Clone, W: Clone> Clone for Condition<H, N, W> {
    fn clone(&self) -> Self {
        Condition {
            now: self.now.clone(),
            wait: self.wait.clone(),
            _pd: PhantomData,
        }
    }
}

impl<H, N, W> TestCondition<H> for Condition<H, N, W>
where
    H: 'static,
    N: 'static,
    W: 'static,
    N: Checker<H>,
    W: Checker<H>,
{
    fn check_now(&self, harness: &H, node: &kdl::KdlNode) -> Result<(), TestErrorCase> {
        let Some(check) = self.now.check(harness) else {
            return Err(TestErrorCase::Error {
                error: miette::miette!("Condition does not implement checking now"),
                label: node.span(),
            });
        };

        check
            .then_some(())
            .ok_or_else(|| crate::error::TestErrorCase::Error {
                error: miette::miette!("Check did not succeed"),
                label: node.span(),
            })
    }

    fn wait_until_true(&self, harness: &H, node: &kdl::KdlNode) -> Result<(), TestErrorCase> {
        let Some(check) = self.wait.check(harness) else {
            return Err(TestErrorCase::Error {
                error: miette::miette!("Condition does not implement waiting"),
                label: node.span(),
            });
        };

        check
            .then_some(())
            .ok_or_else(|| crate::error::TestErrorCase::Error {
                error: miette::miette!("Check did not succeed"),
                label: node.span(),
            })
    }

    fn clone_box(&self) -> Box<dyn TestCondition<H>> {
        Box::new(self.clone())
    }
}
