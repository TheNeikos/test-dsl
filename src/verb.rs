use std::any::Any;
use std::marker::PhantomData;

use crate::arguments::VerbArgument;
use crate::error::TestErrorCase;

pub trait TestVerb<H>: 'static {
    fn run(&self, harness: &mut H, node: &kdl::KdlNode) -> Result<(), TestErrorCase>;
    fn clone_box(&self) -> Box<dyn TestVerb<H>>;
}

impl<H: 'static> Clone for Box<dyn TestVerb<H>> {
    fn clone(&self) -> Self {
        let this: &dyn TestVerb<H> = &**self;
        this.clone_box()
    }
}

pub struct FunctionVerb<H> {
    func: BoxedCallable<H>,
    _pd: PhantomData<fn(H)>,
}

impl<H> Clone for FunctionVerb<H> {
    fn clone(&self) -> Self {
        Self {
            func: self.func.clone(),
            _pd: self._pd,
        }
    }
}

impl<H> FunctionVerb<H> {
    pub fn new<F, T>(func: F) -> Self
    where
        F: CallableVerb<H, T>,
    {
        FunctionVerb {
            func: BoxedCallable::new(func),
            _pd: PhantomData,
        }
    }
}

struct BoxedCallable<H> {
    callable: Box<dyn Any>,
    call_fn: fn(&dyn Any, &mut H, &kdl::KdlNode) -> Result<(), TestErrorCase>,
    clone_fn: fn(&dyn Any) -> Box<dyn Any>,
}

impl<H> Clone for BoxedCallable<H> {
    fn clone(&self) -> Self {
        BoxedCallable {
            callable: (self.clone_fn)(&*self.callable),
            call_fn: self.call_fn,
            clone_fn: self.clone_fn,
        }
    }
}

impl<H> BoxedCallable<H> {
    fn new<F, T>(callable: F) -> Self
    where
        F: CallableVerb<H, T>,
    {
        BoxedCallable {
            callable: Box::new(callable),
            call_fn: |this, harness, node| {
                let this: &F = this.downcast_ref().unwrap();
                this.call(harness, node)
            },
            clone_fn: |this| {
                let this: &F = this.downcast_ref().unwrap();
                Box::new(this.clone())
            },
        }
    }

    fn call(&self, harness: &mut H, node: &kdl::KdlNode) -> Result<(), TestErrorCase> {
        (self.call_fn)(&*self.callable, harness, node)
    }
}

pub trait CallableVerb<H, T>: Clone + 'static {
    fn call(&self, harness: &mut H, node: &kdl::KdlNode) -> Result<(), TestErrorCase>;
}

impl<H, F> CallableVerb<H, ((),)> for F
where
    F: Fn(&mut H) -> miette::Result<()>,
    F: Clone + 'static,
{
    fn call(&self, harness: &mut H, node: &kdl::KdlNode) -> Result<(), TestErrorCase> {
        self(harness).map_err(|error| TestErrorCase::Error {
            error,
            label: node.span(),
        })
    }
}

#[rustfmt::skip]
macro_rules! all_the_tuples {
    ($name:ident) => {
        $name!([], T1);
        $name!([T1], T2);
        $name!([T1, T2], T3);
        $name!([T1, T2, T3], T4);
        $name!([T1, T2, T3, T4], T5);
        $name!([T1, T2, T3, T4, T5], T6);
        $name!([T1, T2, T3, T4, T5, T6], T7);
        $name!([T1, T2, T3, T4, T5, T6, T7], T8);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8], T9);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8, T9], T10);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8, T9, T10], T11);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11], T12);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12], T13);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13], T14);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14], T15);
        $name!([T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15], T16);
    };
}

macro_rules! impl_callable {
    (
        [$($ty:ident),*], $last:ident
    ) => {
        #[allow(non_snake_case, unused_mut)]
        impl<H, F, $($ty,)* $last> CallableVerb<H, ($($ty,)* $last,)> for F
            where
                F: Fn(&mut H, $($ty,)* $last,) -> miette::Result<()>,
                F: Clone + 'static,
                $( $ty: VerbArgument, )*
                $last: VerbArgument,
        {
            fn call(&self, harness: &mut H, node: &kdl::KdlNode) -> Result<(), TestErrorCase> {
                let mut args = node.iter();

                let total_count = 1
                    $(
                        + {
                            const _: () = {
                                #[allow(unused)]
                                let $ty = ();
                            };
                            1
                        }

                    )*;

                let mut running_count = 1;

                $(
                    let arg = args.next().ok_or_else(|| TestErrorCase::MissingArgument {
                        parent: node.span(),
                        missing: format!("This function takes {} arguments, you're missing the {}th argument.", total_count, running_count),
                    })?;

                    let $ty = <$ty as VerbArgument>::from_value(arg).ok_or_else(|| {
                        TestErrorCase::WrongArgumentType {
                            parent: node.name().span(),
                            argument: arg.span(),
                            expected: format!("This function takes a '{}' as its argument here.", <$ty as VerbArgument>::TYPE_NAME),
                        }
                    })?;
                    running_count += 1;
                )*

                let _ = running_count;

                let arg = args.next().ok_or_else(|| TestErrorCase::MissingArgument {
                    parent: node.span(),
                    missing: format!("This function takes {tc} arguments, you're missing the {tc}th argument.", tc = total_count),
                })?;
                let $last = <$last as VerbArgument>::from_value(arg).ok_or_else(|| {
                    TestErrorCase::WrongArgumentType {
                        parent: node.name().span(),
                        argument: arg.span(),
                        expected: format!("This function takes a '{}' as its argument here.", <$last as VerbArgument>::TYPE_NAME),
                    }
                })?;

                self(harness, $($ty,)* $last,).map_err(|error| TestErrorCase::Error {
                    error,
                    label: node.span()
                })
            }
        }
    };
}

all_the_tuples!(impl_callable);

impl<H: 'static> TestVerb<H> for FunctionVerb<H> {
    fn run(&self, harness: &mut H, node: &kdl::KdlNode) -> Result<(), TestErrorCase> {
        let res = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
            self.func.call(harness, node)
        }));

        match res {
            Ok(res) => res,
            Err(error) => {
                let mut message = "Something went wrong".to_string();

                let payload = error;

                if let Some(msg) = payload.downcast_ref::<&str>() {
                    message = msg.to_string();
                }

                if let Some(msg) = payload.downcast_ref::<String>() {
                    message.clone_from(msg);
                }

                Err(TestErrorCase::Panic {
                    error: miette::Error::msg(message),
                    label: node.span(),
                })
            }
        }
    }

    fn clone_box(&self) -> Box<dyn TestVerb<H>> {
        Box::new(self.clone())
    }
}
