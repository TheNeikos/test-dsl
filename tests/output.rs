use std::sync::Arc;

use miette::NamedSource;
use test_dsl::condition::Condition;
use test_dsl::verb::FunctionVerb;

#[test]
fn check_invalid() {
    let ts = test_dsl::TestDsl::<()>::new();

    let tc = ts.parse_document(NamedSource::new(
        "test.kdl",
        Arc::from(
            r#"
            testcase {
            }
            tetcase {
            }
            foobar {
            }
            asd
        "#,
        ),
    ));

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(tc.unwrap_err())));
}

#[test]
fn check_unknown() {
    let ts = test_dsl::TestDsl::<()>::new();

    let tc = ts.parse_document(NamedSource::new(
        "test.kdl",
        Arc::from(
            r#"
            testcase {
                repeat 2 {
                    not_found
                }
            }
        "#,
        ),
    ));

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(tc.unwrap_err())));
}

#[test]
fn check_missing_argument() {
    let ts = test_dsl::TestDsl::<()>::new();

    let tc = ts.parse_document(NamedSource::new(
        "test.kdl",
        Arc::from(
            r#"
            testcase {
                repeat {
                }
            }
        "#,
        ),
    ));

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(tc.unwrap_err())));
}

#[test]
fn check_argument_wrong_type() {
    let ts = test_dsl::TestDsl::<()>::new();

    let tc = ts.parse_document(NamedSource::new(
        "test.kdl",
        Arc::from(
            r#"
            testcase {
                repeat hello {
                }
            }
        "#,
        ),
    ));

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(tc.unwrap_err())));
}

#[test]
fn check_argument_wrong_type_verb() {
    let mut ts = test_dsl::TestDsl::<()>::new();

    ts.add_verb(
        "foobar",
        FunctionVerb::new(|_: &mut (), _: usize| {
            // Nothing
            Ok(())
        }),
    );

    let tc = ts
        .parse_document(NamedSource::new(
            "test.kdl",
            Arc::from(
                r#"
            testcase {
                foobar
            }
        "#,
            ),
        ))
        .unwrap()[0]
        .run(&mut ());

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(tc.unwrap_err())));

    let tc = ts
        .parse_document(NamedSource::new(
            "test.kdl",
            Arc::from(
                r#"
            testcase {
                foobar not_a_number
            }
        "#,
            ),
        ))
        .unwrap()[0]
        .run(&mut ());

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(tc.unwrap_err())));
}

#[test]
fn check_verb_panic_fail() {
    let mut ts = test_dsl::TestDsl::<()>::new();

    ts.add_verb("foobar", FunctionVerb::new(|_: &mut (), _: usize| panic!()));

    let tc = ts
        .parse_document(NamedSource::new(
            "test.kdl",
            Arc::from(
                r#"
            testcase {
                foobar 2 {
                    ofoo
                }
            }
        "#,
            ),
        ))
        .unwrap()[0]
        .run(&mut ());

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(tc.unwrap_err())));
}

#[test]
fn check_conditions() {
    let mut ts = test_dsl::TestDsl::<()>::new();

    ts.add_condition("is_true", Condition::new_now(|_h: &()| Ok(true)));
    ts.add_condition("is_false", Condition::new_now(|_h: &()| Ok(false)));

    let testcases = ts
        .parse_document(NamedSource::new(
            "test.kdl",
            Arc::from(
                r#"
            testcase {
                assert {
                    is_true
                }
            }

            testcase {
                assert {
                    is_false
                }
            }
        "#,
            ),
        ))
        .unwrap();

    // Check that its true
    testcases[0].run(&mut ()).unwrap();

    let is_false = testcases[1].run(&mut ());

    insta::assert_snapshot!(format!("{:?}", miette::Error::new(is_false.unwrap_err())));
}
