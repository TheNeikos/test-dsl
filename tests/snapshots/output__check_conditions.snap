---
source: tests/output.rs
expression: "format!(\"{:?}\", miette::Error::new(is_false.unwrap_err()))"
---
  [31mÃ—[0m Testcase did not run successfully
[31m  â”œâ”€â–¶ [0m  [31mÃ—[0m An error occurred while running the test
[31m  â”‚   [0m    â•­â”€[[36;1;4mtest.kdl:10:21[0m]
[31m  â”‚   [0m [2m 9[0m â”‚                 assert {
[31m  â”‚   [0m [2m10[0m â”‚                     is_false
[31m  â”‚   [0m    Â· [35;1m                    â”€â”€â”€â”€â”¬â”€â”€â”€[0m
[31m  â”‚   [0m    Â·                         [35;1mâ•°â”€â”€ [35;1min this node[0m[0m
[31m  â”‚   [0m [2m11[0m â”‚                 }
[31m  â”‚   [0m    â•°â”€â”€â”€â”€
[31m  â”‚   [0m
[31m  â•°â”€â–¶ [0m  [31mÃ—[0m Assert failed
[31m      [0m
